packages/web3-providers-ws/test/integration/web_socket_provider_integration.test.ts

```diff
@@ -0,0 +1,338 @@
+} from 'web3-common';
+import express from 'express';
+import { createProxyMiddleware } from 'http-proxy-middleware';
+
+import { Server } from 'http';
+import { Web3WSProviderError } from 'web3-errors';
+import WebSocketProvider from '../../src/index';
+// import { Web3WSProviderError } from '../../src/errors';
+import { WSRequestItem, OnCloseEvent } from '../../src/types';
+
+import {
+	getSystemTestProvider,
+	describeIf,
+	getSystemTestAccounts,
+} from '../fixtures/system_test_utils';
+
+describeIf(getSystemTestProvider().includes('ws'))(
+	'WebSocketProvider - implemented methods',
+	() => {
+		let clientWsUrl: string;
+		let accounts: string[];
+		let webSocketProvider: WebSocketProvider;
+		let jsonRpcPayload: Web3APIPayload<EthExecutionAPI, 'eth_getBalance'>;
+		// helper function
+		let currentAttempt = 0;
+		const waitForOpenConnection = async (provider: WebSocketProvider, status = 'connected') => {
+			return new Promise<void>((resolve, reject) => {
+				const maxNumberOfAttempts = 10;
+				const intervalTime = 5000; // ms
+
+				const interval = setInterval(() => {
+					if (currentAttempt > maxNumberOfAttempts - 1) {
+						clearInterval(interval);
+						reject(new Error('Maximum number of attempts exceeded'));
+					} else if (provider.getStatus() === status) {
+						clearInterval(interval);
+						resolve();
+					}
+					// eslint-disable-next-line no-plusplus
+					currentAttempt++;
+				}, intervalTime);
+			});
+		};
+		beforeAll(async () => {
+			clientWsUrl = getSystemTestProvider();
+			accounts = await getSystemTestAccounts();
+		});
+		describe('tests', () => {
+			beforeEach(() => {
+				jsonRpcPayload = {
+					jsonrpc: '2.0',
+					id: 42,
+					method: 'eth_getBalance',
+					params: [accounts[0], 'latest'],
+				} as Web3APIPayload<EthExecutionAPI, 'eth_getBalance'>;
+				webSocketProvider = new WebSocketProvider(
+					clientWsUrl,
+					{},
+					{ delay: 1, autoReconnect: false, maxAttempts: 1 },
+				);
+				currentAttempt = 0;
+			});
+			afterEach(async () => {
+				if (webSocketProvider.getStatus() !== 'disconnected') {
+					// make sure we try to close the connection after it is established
+					await waitForOpenConnection(webSocketProvider);
+					webSocketProvider.disconnect();
+				}
+			});
+
+			describe('websocker provider tests', () => {
+				it('should connect', async () => {
+					await waitForOpenConnection(webSocketProvider);
+					expect(webSocketProvider).toBeInstanceOf(WebSocketProvider);
+					expect(webSocketProvider.getStatus()).toBe('connected');
+				});
+			});
+
+			describe('subscribe event tests', () => {
+				it('should subscribe on message', done => {
+					webSocketProvider.on(
+						'message',
+						(
+							error: Error | null,
+							result?: JsonRpcSubscriptionResult | JsonRpcNotification<any>,
+						) => {
+							if (error) {
+								done.fail(error);
+							}
+							expect(result?.id).toBe(jsonRpcPayload.id);
+							done();
+						},
+					);
```

Comment:
See the usage in contracts test, may be we can convert this callback code to more promise oriented.